Cpalmer/relnotes0.6 (#410)
MicrosoftDocs/quantum-docs-pr: articles/relnotes/index.md

# Microsoft Quantum Development Kit Release Notes

# Version 0.6.1904

*Release date: April 31, 2019 (May 1, 2019)*

This release contains the following:
- makes changes to the Q# language, 
- restructures the Quantum Development Kit libraries, 
- adds a newly contributed sample, and 
- fixes bugs.  

The changes are summarized here as well as instructions for upgrading your existing programs.  You can read more about these changes on devblogs.microsoft.com/qsharp.

## Q# language syntax
This release adds new Q# language syntax:

> * Add a [shorthand way to express variations of quantum operations](xref:microsoft.quantum.language.type-model#functors) (control and adjoints) with "+" operators.  The old syntax is deprecated.  Programs that use the old syntax (e.g., ": adjoint") will continue to work, but a compile time warning will be generated.  

Reviewer: ```suggestion
* Add a [shorthand way to express specializations of quantum operations](xref:microsoft.quantum.language.type-model#functors) (control and adjoints) with `+` operators.  The old syntax is deprecated.  Programs that use the old syntax (e.g., `: adjoint`) will continue to work, but a compile time warning will be generated.  
```

> * Add a new operator for [copy-and-update](xref:microsoft.quantum.language.expressions#copy-and-update-expressions), "w/", can be used to express array creation as a modification of an existing array.

Reviewer: ```suggestion
* Add a new operator for [copy-and-update](xref:microsoft.quantum.language.expressions#copy-and-update-expressions), `w/`, can be used to express array creation as a modification of an existing array.
```

> * Add the common [apply-and-upate statement](xref:microsoft.quantum.qsharp-ref.statements#rebinding-of-mutable-symbols), e.g., "+=", "w/=".

Reviewer: ```suggestion
* Add the common [apply-and-upate statement](xref:microsoft.quantum.qsharp-ref.statements#rebinding-of-mutable-symbols), e.g., `+=`, `w/=`.
```

* Add a way to specify a short name for namespaces in [open directives](xref:microsoft.quantum.language.file-structure#open-directives).

> With this release, we no longer allow an array element to be specified on the left side of a set statement.  This is because that syntax implies that arrays are mutable when in fact, the result of the operation has always been the creation of a new array with the modification.  Instead, a compiler error will be generated with a suggestion to use the new copy-and-update operator, "w/", to accomplish the same result.  

Reviewer: ```suggestion
With this release, we no longer allow an array element to be specified on the left side of a set statement.  This is because that syntax implies that arrays are mutable when in fact, the result of the operation has always been the creation of a new array with the modification.  Instead, a compiler error will be generated with a suggestion to use the new copy-and-update operator, `w/`, to accomplish the same result.  
```

## Library restructuring
This release reorganizes the libraries to enable their growth in a consistent way:

> * Renames namespace Microsoft.Quantum.Primitive  to Microsoft.Quantum.Intrinsic.  These operations are implemented by the target machine.  The Microsoft.Quantum.Primitive namespace is deprecated.  A runtime warning will advise changing the namespace to Intrinsic.

Reviewer: ```suggestion
* Renames the Microsoft.Quantum.Primitive namespace  to Microsoft.Quantum.Intrinsic.  These operations are implemented by the target machine.  The Microsoft.Quantum.Primitive namespace is deprecated.  A runtime warning will advise when programs call operations and functions using deprecated names.
```

> * Renames Microsoft.Quantum.Canon package to Microsoft.Quantum.Standard.  This package contains namespaces that are common to most Q# programs.  This includes:  

Reviewer: ```suggestion
* Renames the Microsoft.Quantum.Canon package to Microsoft.Quantum.Standard.  This package contains namespaces that are common to most Q# programs.  This includes:  
```

    - Microsoft.Quantum.Canon for common operations
    - Microsoft.Quantum.Arithmetic for general purpose arithmetic operations
    - Microsoft.Quantum.Preparation for operations used to prepare qubit state
    - Microsoft.Quantum.Simulation for simulation functionality

With this change, programs that include a single "open" statement for the namespace Microsoft.Quatum.Canon may encounter build errors if the program references operations that were moved to the other three new namespaces.  Adding the additional open statements for the three new namespaces is a straightforward way to resolve this issue.  

* Several namespaces have been deprecated as the operations within have been reorganized to other namespaces. Programs that use these namespaces will continue to work, and a compile time warning will denote the namespace where the operation is defined.  

> * Microsoft.quantum.arithmetic namespace has been normalized to use little endian. Use the function [BigEndianAsLittleEndian](xref:microsoft.quantum.arithmetic.bigendianaslittleendian) when needed to convert to little endian.  

Reviewer: ```suggestion
* The Microsoft.Quantum.Arithmetic namespace has been normalized to use the <xref:microsoft.quantum.arithmetic.littleendian> user-defined type. Use the function [BigEndianAsLittleEndian](xref:microsoft.quantum.arithmetic.bigendianaslittleendian) when needed to convert to little endian.  
```

## Samples and Bug fixes

## Migrating existing projects to 0.6.1904.

  
1.  Follow [these instructions](xref:microsoft.quantum.install#updating-iq) to update IQ#.  
2.  Follow [these instructions](xref:microsoft.quantum.install.python) to update Python.
3.  Follow these instructions to update your .csproj file: 

If you have existing Q# projects from version 0.5 of the Quantum Development Kit, the following are the steps to migrate those projects to the newest version.

1. Projects need to be upgraded in order.  If you have a solution with multiple projects, update each project in the order they are referenced.
2. From a command line, Run `dotnet clean` to remove all existing binaries and intermediate files.
3. In a text editor, edit the .csproj file to change the version of all 
   the "Microsoft.Quantum" `PackageReference` to version 0.6.1904, and change the "Microsoft.Quantum.Canon" package name to "Microsoft.Quantum.Standard", for example:
```xml
    <PackageReference Include="Microsoft.Quantum.Standard" Version="0.6.1904.xxxx" />
    <PackageReference Include="Microsoft.Quantum.Development.Kit" Version="0.6.1904.xxxx" />
```

> 4. From the command line, running this command: `dotnet msbuild`  

Reviewer: ```suggestion
4. From the command line, run this command: `dotnet msbuild`  
```
